### ISSUE TITLE
DisallowedMethods cop?

### ISSUE THREAD

**Issue by mockdeep** [rubocop/rubocop#1065]
Kind of similar to the `CollectionMethods` cop, it would be nice to have a cop that disallows certain methods. For example, we're switching over our spec syntax from:

``` ruby
User.any_instance.should_receive(:foo).and_return('blah')
```

to

``` ruby
expect_any_instance_of(User).to receive(:foo).and_return('blah')
```

It would be nice to ban `should_receive` calls from our codebase.

**Comment by TSMMark** [comment]
Would like to disallow `abort` calls (Kernel, Process, etc) from codebase (and prefer `raise` instead). Is there a rule for that, or would this closed issue be the closest thing? Could this be reopened?

**Comment by dvandersluis** [comment]
That’s probably a great candidate for a custom cop, and should be pretty simple to output (create a cop with an `on_send` method and `RESTRICT_ON_SEND = %i[abort]` and you’re 80% there). 

Like @bbatsov said having a generic cop for replacing method x with method y is problematic because all context is lost. Of course you could probably add a custom cop to do so too but having more specific cops is beneficial IMO.

**Comment by TSMMark** [comment]
Wrote a custom cop, here it is for anyone else that comes across this issue and wants a reference how to do it. 

cop:

```ruby
require "rubocop"

module RuboCop
  module Cop
    module Style
      class DisallowMethods < Cop

        # Feel free to add additional disallowed methods here and update the MSG along with the code.
        MSG = "You should use `raise` instead of `abort`".freeze
        DISALLOWED_METHODS = %i[
          abort
        ].freeze

        def on_send(node)
          return unless DISALLOWED_METHODS.include?(node.method_name)

          add_offense(node)
        end

        def autocorrect(node)
          lambda do |corrector|
            case node.method_name
            when :abort then corrector.replace(node.loc.selector, "raise")
            end
          end
        end

      end
    end
  end
end
```

spec:

```ruby
require_relative "../spec_helper"

describe :"RuboCop::Cop::Style::DisallowMethods" do
  include RuboCopCopSpec

  let(:cop_class) { RuboCop::Cop::Style::DisallowMethods }

  it "registers offense when abort is used" do
    assert_offense(<<~RUBY)
      def validate
        abort "Whatever" if some_condition?
        ^^^^^^^^^^^^^^^^ You should use `raise` instead of `abort`
        anything
      end
    RUBY

    assert_correction(<<~RUBY)
      def validate
        raise "Whatever" if some_condition?
        anything
      end
    RUBY
  end

  it "does not register offense when raise is used" do
    assert_no_offenses(<<~RUBY)
      def validate
        raise "Whatever" if some_condition?
        anything
      end
    RUBY
  end
end
```

**Comment by ndbroadbent** [comment]
I found this issue while searching for a way to disable calls to `puts` in our codebase (to prevent debugging lines from being accidentally committed.)

`rubocop-rails` has a `Rails/Output` cop for this: https://www.rubydoc.info/gems/rubocop-rails/RuboCop/Cop/Rails/Output

**Comment by mockdeep** [comment]
Wow, I totally forgot about this issue. Almost 10 years later and barely a week goes by without me wishing I had something like this. There are always methods that I would like to deprecate or discourage in my codebases, but it doesn't make sense to have a linter for each of them. E.g.:

* Don't call `.click` or `click_link` or `click_button` in Capybara specs, use `click_on`
* Don't call `form_tag` or `form_for`, use `form_with`
* Don't call the deprecated `assigns` or `render_template` in controller specs, assert against view contents

Every case I can think of the method is unique, so there isn't really a context issue, but it would be nice to have a description to explain alternatives. I'll play around with the custom linter by @TSMMark.

**Comment by vanyauhalin** [comment]
@mockdeep, could you please consider opening this issue to prevent it from being lost again? I believe that this feature is still required by many. 

Personally, I would like to BAN all standard module extensions that Rails includes. 🌚

**Comment by mockdeep** [comment]
@vanyauhalin I don't have permissions to reopen issues. @bbatsov would be able to, though.

**Comment by TSMMark** [comment]
In the few years since originally posting the custom linter above, we've made some improvements. Since there's still some traffic on this issue, I'll post the most recent iteration:

 - Updated to the new rubocop rule authoring patterns.
 - Can now autocorrect to replace certain methods with others e.g. `ends_with?` -> `end_with?` for consistency
 - Uses `RESTRICT_ON_SEND` for rubocop perf

```ruby
require "rubocop"

module RuboCop
  module Cop
    module Style
      class DisallowMethods < Base

        extend AutoCorrector

        # Feel free to add additional disallowed methods here and update the MSG along with the code.
        DIFF_METH_MSG = "You should use `%<new>s` instead of `%<old>s`".freeze
        NO_METH_MSG = "You should not use `%<old>s` %<desc>s".freeze
        DIFF_METHS = {
          abort: "raise",
          ends_with?: "end_with?",
          includes?: "include?",
          starts_with?: "start_with?",
        }.freeze
        NO_METHS = {
          snail!: "except while testing locally",
        }.freeze

        RESTRICT_ON_SEND = (DIFF_METHS.keys + NO_METHS.keys).freeze

        def on_send(node)
          name = node.method_name
          if DIFF_METHS.key?(name)
            add_offense(node, message: format(DIFF_METH_MSG, old: name, new: DIFF_METHS[name])) { |corrector| autocorrect(corrector, node) }
          elsif NO_METHS.key?(name)
            add_offense(node, message: format(NO_METH_MSG, old: name, desc: NO_METHS[name])) { |corrector| autocorrect(corrector, node) }
          end
        end

        def autocorrect(corrector, node)
          name = node.method_name
          case name
          when *DIFF_METHS.keys then corrector.replace(node.loc.selector, DIFF_METHS[name])
          when *NO_METHS.keys then corrector.replace(node.loc.selector, "")
          else
            raise "Unexpected method name #{name.inspect}"
          end
        end

      end
    end
  end
end
```
